### PR TITLE
Make it build with ghc-9.10

### PR DESCRIPTION
**Pull request by erikd**
# Description

Make it build with ghc-9.10.

Cannot be merged until after the SRPs have been removed.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

**Comment by neilmayhew** [comment]
I think it would be better to make all uses of `foldl'` from `Foldable` unqualified, and where there's an ambiguity with one of the specialized versions (eg from `Set`) make those qualified.

Even with ghc 8.10.7, `Data.List.foldl'` is just a reexport of `Data.Foldable.foldl'`, so I think we should also change all the uses of `Data.List.foldl'` to `Data.Foldable.foldl'`.

**Comment by lehins** [comment]
> I think it would be better to make all uses of foldl' from Foldable unqualified, and where there's an ambiguity with one of the specialized versions (eg from Set) make those qualified.

@neilmayhew The reason for qualified import is because GHC-9.10 finally exposes `foldl'` from `Prelude`, which means we run into redundant constraint issue for older ghc. So, the cleanest way to solve this is by a qualified import of `Foldable`

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs [hunk 1]
```diff
@@ -100,7 +100,7 @@ import Control.State.Transition.Extended
 import qualified Data.ByteString.Lazy as BSL (length)
 import Data.Coerce (coerce)
 import Data.Either (isRight)
-import Data.Foldable (foldl', sequenceA_, toList)
+import Data.Foldable as Foldable (foldl', sequenceA_, toList)
```
Comment (lehins) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs]: Can you please rename all the qualifications in this PR from `Foldable` to `F`. This is needed for consistency, since that is what we use in other places where `Foldable` in imported qualified
```suggestion
import Data.Foldable as F (foldl', sequenceA_, toList)
```

File: eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Utxo.hs [hunk 2]
```diff
@@ -63,7 +63,7 @@ import Control.State.Transition
 import qualified Data.ByteString.Lazy as BSL
 import qualified Data.Either as Either (partitionEithers)
 import qualified Data.IntSet as IntSet
-import Data.List (foldl')
+import Data.List as List (foldl')
```
Comment (lehins) [inline eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Utxo.hs]: For consistency sake again, we should start using `Foldable`: 
```suggestion
import Data.Foldable as F (foldl')
```

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs (hunk 1 above)
Comment (erikd) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs]: Done

File: eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Utxo.hs (hunk 2 above)
Comment (erikd) [reply eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Utxo.hs]: Done

**Comment by erikd** [comment]
I am not worried about the nix eval failure for now. I will fix it when the rest works.

File: libs/cardano-ledger-core/cardano-ledger-core.cabal [hunk 3]
```diff
@@ -113,7 +113,7 @@ library
         measures,
         microlens,
         network,
-        nothunks,
+        nothunks ^>=0.1.5,
```
Comment (erikd) [inline libs/cardano-ledger-core/cardano-ledger-core.cabal:116]: Version after this has an API change. Tried to add a comment but then `cabal-format` complained.

File: cabal.project [hunk 4]
```diff
@@ -22,16 +22,18 @@ source-repository-package
...
+  , cardano-haskell-packages 2024-07-23T08:58:13Z
+
 
 packages:
   eras/allegra/impl
   eras/alonzo/impl
   eras/alonzo/test-suite
   eras/babbage/impl
   eras/babbage/test-suite
+  eras/byron/crypto
```
Comment (teodanciu) [inline cabal.project]: This is already mentioned in this list, a [few lines below](https://github.com/IntersectMBO/cardano-ledger/pull/4451/files#diff-b8ed06757045fac949c89f2139a862498ad2b6d1f82c61a31e7c91d6cf0eaa70L61)

Is it on purpose  duplicated here?

File: eras/byron/crypto/cardano-crypto-wrapper.cabal [hunk 5]
```diff
@@ -83,7 +83,7 @@ library
         canonical-json,
         cardano-ledger-binary >=1.3.1,
         cardano-crypto,
-        cardano-prelude >=0.1.0.1,
+        cardano-prelude >=0.2.0.0,
```
Comment (teodanciu) [inline eras/byron/crypto/cardano-crypto-wrapper.cabal:86]: Could you please increment the package version in this cabal file to `1.5.1.3` ?

File: libs/cardano-ledger-binary/cardano-ledger-binary.cabal [hunk 6]
```diff
@@ -67,7 +67,7 @@ library
         network,
         nothunks,
         primitive,
-        plutus-ledger-api >=1.27.0 && <1.31.0,
+        plutus-ledger-api ^>=1.31,
```
Comment (teodanciu) [inline libs/cardano-ledger-binary/cardano-ledger-binary.cabal]: This also requires incrementing the package version to `1.3.3.1`

**Review by teodanciu — APPROVED** [review]
Looks good to me, some minor version updates and a seemingly unnecessary duplicated package declaration in cabal file.

File: cabal.project (hunk 4 above)
Comment (lehins) [reply cabal.project]:
```suggestion
```

File: cabal.project [hunk 7]
```diff
@@ -96,3 +98,17 @@ benchmarks: true
 
 -- The only sensible test display option
 test-show-details: streaming
+
+
+constraints:
+  -- The API has changed for version 2.2, and there is no intention to change to that
+  -- version until after Conway.
+  , cardano-crypto-class ^>= 2.1
```
Comment (lehins) [inline cabal.project]: Why is this necessary? API have changed in `cardano-crypto-class` current version on `master`, but the version 2.2 has not even been released yet, so this kind of speculative bounds should not exist in any `cabal.project` file.
```suggestion
```

File: libs/cardano-ledger-binary/cardano-ledger-binary.cabal (hunk 6 above)
Comment (lehins) [reply libs/cardano-ledger-binary/cardano-ledger-binary.cabal]: Please, don't touch the lower bound when it is specified explicitely, unless you are sure that there is an incompatibility. In this case there is no issues with compatibility, so only upper bound needs adjustment
```suggestion
        plutus-ledger-api >=1.27.0 && <1.32.0,
```

File: libs/cardano-ledger-core/cardano-ledger-core.cabal (hunk 3 above)
Comment (lehins) [reply libs/cardano-ledger-core/cardano-ledger-core.cabal:116]: Could you please post here the build error that we get when using `nothunks-0.2.x` (i.e. without this constraint)?

If we can't accommodate it in this PR then we'll be able to turn this comment into its own issue and take care of it later.

File: cabal.project (hunk 7 above)
Comment (erikd) [reply cabal.project]: Removed. Pretty certain I added that when I was using an SRP for `cardano-base`.

File: libs/cardano-ledger-core/cardano-ledger-core.cabal (hunk 3 above)
Comment (erikd) [reply libs/cardano-ledger-core/cardano-ledger-core.cabal:116]: https://github.com/IntersectMBO/cardano-ledger/issues/4504
